sw/device/tests/rv_core_ibex_rnd_test.c

```diff
@@ -0,0 +1,93 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_edn.h"
+#include "sw/device/lib/dif/dif_rv_core_ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/entropy_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "rv_core_ibex_regs.h"
+
+// Initialize OTTF
+OTTF_DEFINE_TEST_CONFIG();
+
+// Declare two assembly function defined in rv_core_ibex_rnd_test.S
+extern volatile uint32_t load_from_two_addresses(volatile uint32_t *address1,
+                                                 volatile uint32_t *address2);
+extern volatile uint32_t interleave_loads_from_two_addresses(
+    volatile uint32_t *address1, volatile uint32_t *address2);
+
+#define RANDOM_DATA_READS 32
+#define MAX_STATUS_CHECKS 1024
+
+static void wait_till_status_is_valid(const dif_rv_core_ibex_t *rv_core_ibex) {
+  int j = 0;
+  dif_rv_core_ibex_rnd_status_t rnd_status = 0;
+  CHECK(rv_core_ibex != NULL);
+
+  while (rnd_status == 0) {
+    CHECK_DIF_OK(dif_rv_core_ibex_get_rnd_status(rv_core_ibex, &rnd_status));
+    j += 1;
+    if (j == MAX_STATUS_CHECKS) {
+      // Make sure that this while loop cannot last infinitely
+      CHECK(false);
+    }
+  }
+}
+
+bool test_main(void) {
+  // Verify the functionality of the random number generation CSRs.
+
+  // Enable entropy complex, CSRNG and EDN so Ibex can get entropy.
+  entropy_testutils_boot_mode_init();
+
+  // Initialize Ibex.
+  dif_rv_core_ibex_t rv_core_ibex;
+  CHECK_DIF_OK(dif_rv_core_ibex_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR),
+      &rv_core_ibex));
+
+  // Perform multiple reads from `RND_DATA` polling `RND_STATUS` in between to
+  // only read valid data. Check different random bits are provided each time.
+  uint32_t rnd_data;
+  uint32_t previous_rnd_data = 0;
+  for (int i = 0; i < RANDOM_DATA_READS; i++) {
+    wait_till_status_is_valid(&rv_core_ibex);
+    CHECK_DIF_OK(dif_rv_core_ibex_read_rnd_data(&rv_core_ibex, &rnd_data));
+    if (i != 0) {
+      CHECK(rnd_data != previous_rnd_data);
+    }
+    previous_rnd_data = rnd_data;
+  }
+
+  // Ensure `RND_STATUS` indicate invalid data immediately after `RND_DATA`
+  // read.
+  volatile uint32_t *rnd_data_address =
+      &((volatile uint32_t *)rv_core_ibex.base_addr
+            .base)[RV_CORE_IBEX_RND_DATA_REG_OFFSET / sizeof(uint32_t)];
+  volatile uint32_t *rnd_status_address =
+      &((volatile uint32_t *)rv_core_ibex.base_addr
+            .base)[RV_CORE_IBEX_RND_STATUS_REG_OFFSET / sizeof(uint32_t)];
+  CHECK(load_from_two_addresses(rnd_data_address, rnd_status_address) == 0);
```

Comment:
I would make the `load_from_two_addresses` even more specific and make it access the register macros `RV_CORE_IBEX_RND_DATA_REG_OFFSET` directly and check the invalid status.
Then you could name it something like `rv_core_ibex_rnd_read_and_immediately_check_status`.